### PR TITLE
Replace rmm::device_vector & thrust::host_vector with rmm::device_uvector & std::vector, respectively.

### PR DESCRIPTION
**Pull request by seunghwak**
- [x] Replace rmm::device_vector with rmm::device_uvector for better concurrency in multi-stream executions
- [x] Replace thrust::host_vector with std::vector

This PR partially addresses https://github.com/rapidsai/cugraph/issues/1390

### CODE REVIEW

**Comment by seunghwak** [comment]
@afender you may want to review this.

File: cpp/include/patterns/vertex_frontier.cuh [hunk 1]
```diff
@@ -147,13 +147,17 @@ template <typename vertex_t, bool is_multi_gpu = false>
 class Bucket {
  public:
   Bucket(raft::handle_t const& handle, size_t capacity)
-    : handle_ptr_(&handle), elements_(capacity, invalid_vertex_id<vertex_t>::value)
+    : handle_ptr_(&handle), elements_(capacity, handle.get_stream())
   {
+    thrust::fill(rmm::exec_policy(handle_ptr_->get_stream())->on(handle_ptr_->get_stream()),
```
Comment (afender) [inline cpp/include/patterns/vertex_frontier.cuh:152]: Consider replacing 
`thrust::fill(rmm::exec_policy(handle_ptr_->get_stream())->on(handle_ptr_->get_stream()),`
by 
`thrust::fill(rmm::exec_policy(handle.get_stream_view()),`

This applies to all thrust calls

**Review by afender — COMMENTED** [review]
Looks good but :
- consider using the latest streams features from RAFT
- could we think of a test to verify it is now executing on the stream?

File: cpp/include/compute_partition.cuh [hunk 2]
```diff
@@ -39,27 +39,33 @@ class compute_partition_t {
...
 
-    vertex_partition_offsets_v_.resize(size_ + 1);
-    vertex_partition_offsets_v_ = partition.get_vertex_partition_offsets();
+    vertex_partition_offsets_v_.resize(size_ + 1, handle.get_stream());
+    auto vertex_partition_offsets = partition.get_vertex_partition_offsets();
+    raft::update_device(vertex_partition_offsets_v_.data(),
+                        vertex_partition_offsets.data(),
+                        vertex_partition_offsets.size(),
+                        handle.get_stream());
+    CUDA_TRY(cudaStreamSynchronize(handle.get_stream()));
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/compute_partition.cuh:68]: Oops, no :-)

File: cpp/include/patterns/count_if_e.cuh [hunk 3]
```diff
@@ -201,7 +201,7 @@ typename GraphViewType::edge_type count_if_e(
                                          detail::count_if_e_for_all_block_size,
                                          handle.get_device_properties().maxGridSize[0]);
 
-      rmm::device_vector<edge_t> block_counts(update_grid.num_blocks);
+      rmm::device_uvector<edge_t> block_counts(update_grid.num_blocks, handle.get_stream());
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/patterns/count_if_e.cuh:204]: Yeah... get_stream() is used all around the codebase. I will replace all these at once after multiple pending PRs get merged.

File: cpp/include/patterns/update_frontier_v_push_if_out_nbr.cuh [hunk 4]
```diff
@@ -673,15 +671,19 @@ void update_frontier_v_push_if_out_nbr(
...
 
     auto bucket_sizes_device_ptr = std::get<1>(bucket_and_bucket_size_device_ptrs);
-    thrust::host_vector<size_t> bucket_sizes(
-      bucket_sizes_device_ptr, bucket_sizes_device_ptr + VertexFrontierType::kNumBuckets);
+    std::vector<size_t> bucket_sizes(VertexFrontierType::kNumBuckets);
+    raft::update_host(bucket_sizes.data(),
+                      bucket_sizes_device_ptr,
+                      VertexFrontierType::kNumBuckets,
+                      handle.get_stream());
+    CUDA_TRY(cudaStreamSynchronize(handle.get_stream()));
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/patterns/update_frontier_v_push_if_out_nbr.cuh:686]: This is necessary as without this, bucket_sizes (which is std::vector and is not aware of stream) is not guaranteed to hold up-to-date data and is used right after this sync.

```CUDA_TRY(cudaStreamSynchronize(handle.get_stream()));``` this is also used all around the codebase, so I will create an issue to replace all at once.

File: cpp/include/patterns/vertex_frontier.cuh (hunk 1 above)
Comment (seunghwak) [reply cpp/include/patterns/vertex_frontier.cuh:152]: Yes, same story here. This should be replaced all at once.

File: cpp/include/patterns/count_if_e.cuh (hunk 3 above)
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/patterns/count_if_e.cuh:204]: And with get_stream_view(), do we need get_stream()? Should we better deprecate (and eventually remove) raft::handle_t's get_stream()?

**Comment by seunghwak** [comment]
@afender https://github.com/rapidsai/cugraph/issues/1422 is created to transition cuGraph to rmm's new cuda_stream_view.

File: cpp/include/compute_partition.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/include/compute_partition.cuh]: Removed

**Comment by seunghwak** [comment]
> * could we think of a test to verify it is now executing on the stream?

This is tricky I guess; the first thing is this is not a black-or-white thing and this will affect only performance not correctness.

I guess we need cuGraph-wise benchmark scheme to track multi-stream execution of (every) graph analytics?

File: cpp/include/patterns/count_if_e.cuh (hunk 3 above)
[the remark this answers is not in the document]
Comment (afender) [reply cpp/include/patterns/count_if_e.cuh:204]: Indeed, but it is used everywhere in cugraph and cuml so we'll need to get rid of these occurrences first.

**Review by afender — APPROVED** [review]

**Comment by afender** [comment]
> > * could we think of a test to verify it is now executing on the stream?
> 
> This is tricky I guess; the first thing is this is not a black-or-white thing and this will affect only performance not correctness.
> 
> I guess we need cuGraph-wise benchmark scheme to track multi-stream execution of (every) graph analytics?

Ok I can rebase my concurrency branch and generate a profile after this gets merged.

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**